### PR TITLE
RA-2023: Making configs validated through Iniz Validator.

### PR DESCRIPTION
**Pull request by Ruhanga**
Ticket: https://issues.openmrs.org/browse/RA-2023
Depends on https://github.com/openmrs/openmrs-contrib-packager-maven-plugin/pull/7

### CODE REVIEW

File: distro/pom.xml
Comment (ibacher) [inline distro/pom.xml:189]: We need to do a release of the Maven plugin before this can be incorporated. I think it's OK to tolerate SNAPSHOTs in the properties that are driving the distro configuration, but completely wrong to do that for Maven plugins.

File: distro/configuration/concepts/misc-core_data.csv
Comment (ibacher) [inline distro/configuration/concepts/misc-core_data.csv:2]: Why are we removing these concepts?

File: distro/configuration/concepts/drugs_concepts.csv
Comment (ibacher) [inline distro/configuration/concepts/drugs_concepts.csv:1]: Ditto here as for drugs: the concepts need to be fixed. We shouldn't be removing them.

**Review by ibacher — COMMENTED** [review]
Thanks @Ruhanga! This is important work, but it's also important that we not delete metadata just because it fails in the validator. We need to either fix the metadata or fix the validator.

File: distro/pom.xml
Comment (Ruhanga) [reply distro/pom.xml:189]: Yes for sure. That's precisely the plan. I'd simply done this in awaiting a release of the plugin.

File: distro/configuration/metadatatermmappings/metadatatermmappings.csv
Comment (Ruhanga) [reply distro/configuration/metadatatermmappings/metadatatermmappings.csv:1]: For a while these don't seem to be necessarily needed (since no one is noticing their absence in a Ref App distro) and so IMO, removing them won't do any harm I guess.

File: distro/configuration/concepts/misc-core_data.csv
Comment (Ruhanga) [reply distro/configuration/concepts/misc-core_data.csv:2]: These are duplicates and already come with the OCL packages.

File: distro/configuration/concepts/drugs_concepts.csv
Comment (Ruhanga) [reply distro/configuration/concepts/drugs_concepts.csv:1]: Removing them sounds like the right fix since they are duplicates already brought in by the OCL packages.

File: distro/configuration/drugs/drugs.csv
Comment (Ruhanga) [reply distro/configuration/drugs/drugs.csv:23]: For a while they don't load and this is not noticed by anyone needing them. IMO I'd rather remove them so they are actually added when needed.

File: distro/configuration/concepts/drugs_concepts.csv
Comment (ibacher) [reply distro/configuration/concepts/drugs_concepts.csv:1]: Ok... As long as their all duplicated.

File: distro/configuration/metadatatermmappings/metadatatermmappings.csv
Comment (Ruhanga) [reply distro/configuration/metadatatermmappings/metadatatermmappings.csv:1]: @ibacher, I've been looking into this silent failure to load the `emrapi.Metadata_Source-1.zip` and the PR below should fix it.

https://github.com/mekomsolutions/openmrs-module-initializer/pull/241

**Review by Ruhanga — COMMENTED** [review]
Hi @ibacher, could you review and share your final remarks? Thanks. I've added github actions too. See the branch's successful build [here](https://github.com/Ruhanga/openmrs-distro-referenceapplication/tree/validation).

File: distro/pom.xml
Comment (Ruhanga) [reply distro/pom.xml]: Done. The Plugin used is 1.7.0.

**Comment by gracepotma** [comment]
@Ruhanga I'm concerned about the metadata deletions. Can you tell me why these drug concepts are being removed? 
If the problem is being caused due to those line items being empty in the strength field (e.g. not having any mg's assigned) I can fix that. Since I introduced that.

**Comment by Ruhanga** [comment]
@gracepotma, at the moment the deleted drugs are as good as not present  (even before deletion). They are already failing. I would clean them up in favour of adding them on an error free configuration.

**Comment by Ruhanga** [comment]
>Would you like me to, soon, go through all drugs and ensure those are in /ocl as well, for consistency?

@gracepotma you could do so at-least soon after this is merged and released?

**Comment by gracepotma** [comment]
Ah gotcha, thank you Nathan. That totally makes sense then. And sure, I've ticketed this for myself here: https://issues.openmrs.org/browse/O3-2077

File: .github/workflows/build-test.yml
Comment (ibacher) [inline .github/workflows/build-test.yml:5]:
```suggestion
    branches: [main]
```

File: .github/workflows/build-test.yml
Comment (ibacher) [inline .github/workflows/build-test.yml:21]:
```suggestion
        uses: actions/setup-java@v3
```

File: .github/workflows/build-test.yml
Comment (ibacher) [inline .github/workflows/build-test.yml:24]:
```suggestion
          java-version: ${{ matrix.java }}
          cache: 'maven'
```

File: pom.xml
Comment (ibacher) [inline pom.xml:169]: I don't think we need this change; specifically, `public` aliases all the repos in the OpenMRS Maven instance; `releases` does not.

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks @Ruhanga! I think we need some adjustments to the POM changes and the GitHub Actions, but otherwise, I'm happy with this.

File: .github/workflows/build-test.yml
Comment (ibacher) [reply .github/workflows/build-test.yml:5]: We use `main` not `master` in this repo.

File: .github/workflows/build-test.yml
Comment (ibacher) [reply .github/workflows/build-test.yml:8]: We need to exclude the branches for other versions. Tbh, I'd probably just drop the `push` trigger altogether.

File: pom.xml
Comment (Ruhanga) [reply pom.xml:119]: Something weird's happening the OpenMRS repo. If an artefact is missing or poor redirection happens, builds fail. That's why I'm leaving it for last to prevent invalid failures.

File: pom.xml
Comment (Ruhanga) [reply pom.xml:169]: This is somewhat connected to [this](https://openmrs.slack.com/archives/C03U2P841/p1682617009756369?thread_ts=1682419613.318009&cid=C03U2P841) suggestion @ibacher.

Comment (Ruhanga) [reply pom.xml:169]: So what's happening is the openmrs packager plugin artefact isn't downloadable via `https://mavenrepo.openmrs.org/public`, but is via `https://openmrs.jfrog.io/artifactory/releases`. Could this be a problem with the alias url or jfrog?

Comment (Ruhanga) [reply pom.xml]: I think I get your point @ibacher. In  essence I should use `https://openmrs.jfrog.io/artifactory/public` instead of `https://openmrs.jfrog.io/artifactory/releases`?

File: pom.xml
Comment (mks-d) [reply pom.xml:119]: @ibacher in the end something's fishy when there's redirections to another Maven repository. In particular this happens with Iniz because it is proxied maybe? And it looks like letting the build fetch it from the original repo first solves the issue.

In short this proxying is just not working really, we should never have had to add Mekom's repos here, let alone first.

So while I agree that this is not idea altogether, it seems to finally unblock the build.

Comment (Ruhanga) [reply pom.xml:119]: @ibacher, where you able to disable proxying for the Mekom repo on the OpenMRS repos so I could try again re-arranging the repositories?

Comment (ibacher) [reply pom.xml:119]: We aren't proxying the Mekom repo in the OpenMRS repo any more... I'm not sure that we have been for a while. What's weird about this to me is that we've been successfully pulling down, e.g., Iniz with the previous ordering.

Comment (ibacher) [reply pom.xml]: So when I try to run this PR the error I see is this:

```
Execution validate-configurations of goal org.openmrs.maven.plugins:openmrs-packager-maven-plugin:1.7.0:validate-configurations failed: Plugin org.openmrs.maven.plugins:openmrs-packager-maven-plugin:1.7.0 or one of its dependencies could not be resolved: Could not find artifact org.openmrs.module:initializer-validator:jar:2.5.2 in openmrs-repo (https://openmrs.jfrog.io/artifactory/public)
```

There are two entries in the POM, both labelled `openmrs-repo`, one in repositories and one in pluginRepositories. We are able to find Initializer when included as a dependency (it is a dependency of the `distro` sub-project and this works), but we fail to find it when it is a dependency of a Maven plugin. I'm guessing this is because we have a separate pluginRepositories element, but with no reference to the Mekom repos there, which is why we're getting the unresolved thing. This has nothing to do with repo order or proxying or anything like this. We've just told Maven to only search two repositories for plugins and dependencies, and neither of them have Initializer in them.

Adding the Mekom Nexus references to the pluginRepositories element resolves the issue, at least for me.

Comment (rbuisson) [reply pom.xml]: What's weird here is that when running the build in Gitpod, it really works fine.

It's able to fetch the `initializer-validator 2.5.2` as it should (even with the Mekom repos being under the OpenMRS ones, as well as with using `https://mavenrepo.openmrs.org/releases` instead of `https://openmrs.jfrog.io/artifactory/releases`)

Comment (ibacher) [reply pom.xml]: I might guess that your Gitpod instance had already separately resolved initializer-validator 2.5.2. If the dependency (especially a released dependency) is in the local Maven cache, I don't think Maven worries too much about where it was resolved from.

**Comment by rbuisson** [comment]
Thanks for your last fix @ibacher. Seems to fix the CI issue, are we good to merge now?

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
I think so, yeah

**Pull request merged**
